### PR TITLE
Remove SDL2 dependencies: Phase 0

### PR DESCRIPTION
**Pull request by IntelOrca**
These are some changes that I made as part of the movement to a core OpenRCT2 library that can be compiled without `SDL2` and therefore allow servers to more easily build a headless version of OpenRCT2.

These changes replace various SDL2 calls with C or STL calls instead. The HTTP and TcpSocket interfaces can be improved at a later date to take advantage of C++ promises and futures.

### CODE REVIEW

**Comment by janisozaur** [comment]
There's something wrong with the clock now, e.g. the chat messages never time out.

**Comment by janisozaur** [comment]
I will need to switch mingw to use Arch as their base instead of Ubuntu for this PR, it may take a while.

File: src/openrct2/platform/shared.c [hunk 1]
```diff
@@ -763,7 +765,21 @@ void platform_set_cursor_position(sint32 x, sint32 y)
...
+        log_fatal("clock_gettime failed");
+        exit(-1);
+    }
+    return (uint32)(ts.tv_nsec / 1000000);
+#endif
+}
+
+void platform_sleep(uint32 ms)
+{
+    SDL_Delay(ms);
```
Comment (janisozaur) [inline src/openrct2/platform/shared.c]: This is introducing new code depending on SDL. Is that meant to be that way?

Comment (IntelOrca) [inline src/openrct2/platform/shared.c]: For Windows you can call `Sleep(ms`).
I haven't worked out how to do this yet for unix.

Comment (LRFLEW) [inline src/openrct2/platform/shared.c]: If we're using C++11 and newer, we can just use `std::this_thread::sleep_for` 😜

Comment (IntelOrca) [inline src/openrct2/platform/shared.c]: Yea ok, although I might call `Sleep` conditionally for Windows as its much cheaper than the code generated by `sleep_for`.

http://stackoverflow.com/questions/32904371/windows-7-overshoot-c-stdthis-threadsleep-for/32963363

**Comment by ensiform** [comment]
Couldn't/Shouldn't SDL2 be used for Clients and headless version would use alternative or not at all?

This is how other games I have seen accomplish this. By only linking SDL2 stuff to the rendered version/"client". But leaving the headless/console build without.

**Comment by duncanspumpkin** [comment]
@ensiform thats exactly the point of this. Its not simple to do though.

**Comment by ensiform** [comment]
@duncanspumpkin I see. When I first looked at this I thought you would be removing it entirely.

**Comment by duncanspumpkin** [comment]
@ensiform its Phase 0 of many. Its not a simple process to detangle everything.

**Comment by janisozaur** [comment]
@ensiform we aim to have sdl-less builds. Whether this is client or server (in context of multiplayer) is irrelevant.

**Comment by janisozaur** [comment]
Mingw should now be able to build this, I will restart the travis run, but ideally it should be rebased

**Comment by janisozaur** [comment]
@LRFLEW @marijnvdwerf can you fix xcode project?

**Comment by marijnvdwerf** [comment]
```
/Users/Marijn/Projects/OpenRCT2/Main/src/openrct2/core/FileStream.hpp:120:16: Use of undeclared identifier 'ftello64'; did you mean 'ftello'?
/Users/Marijn/Projects/OpenRCT2/Main/src/openrct2/core/FileStream.hpp:146:13: Use of undeclared identifier 'fseeko64'; did you mean 'fseeko'?
/Users/Marijn/Projects/OpenRCT2/Main/src/openrct2/core/FileStream.hpp:149:13: Use of undeclared identifier 'fseeko64'; did you mean 'fseeko'?
/Users/Marijn/Projects/OpenRCT2/Main/src/openrct2/core/FileStream.hpp:152:13: Use of undeclared identifier 'fseeko64'; did you mean 'fseeko'?
```

**Comment by marijnvdwerf** [comment]
```
src/openrct2/network/TcpSocket.cpp:90:10: 'future<void>' is unavailable: introduced in macOS 10.8
src/openrct2/network/TcpSocket.cpp:322:14: 'promise<void>' is unavailable: introduced in macOS 10.8
src/openrct2/network/TcpSocket.cpp:324:63: 'promise<void>' is unavailable: introduced in macOS 10.8
```

**Comment by marijnvdwerf** [comment]
`clock_gettime` was only introduced in macOS 10.12 (the most recent version)

**Comment by janisozaur** [comment]
@marijnvdwerf what version of osx do we target now? Can we target 10.8?

**Comment by marijnvdwerf** [comment]
Now we target 10.7. I don't think increasing it to 10.8 will lose that many people.

**Comment by AaronVanGeffen** [comment]
Agreed with increasing to 10.8. For what it's worth, Apple itself has dropped support for anything older than macOS 10.10.

**Comment by janisozaur** [comment]
@AaronVanGeffen @marijnvdwerf can you update it then?

**Comment by LRFLEW** [comment]
For the record, (last I checked) we decided that we would officially support the last three major versions of macOS (the same way Homebrew does) and provide "best effort" for everything else. Right now, it compiles and runs on 10.7 (There's an issue involving SSL and dependencies on 10.7 and 10.8, but otherwise it works perfectly fine). If we want/need to make a code change that increases the minimum version, then ok. It would be a good idea to recompile the dependencies to match that minimum version (so that CPU extensions and other stuff can be properly optimized), but as long as we support the last 3 major versions (currently 10.10 to 10.12), we should keep most of our audience. (Though there are a surprising number of people still using 10.6 and 10.7).

**Comment by janisozaur** [comment]
I was hoping we would update the target to 10.8, but it sounds like it may be better to just go directly to 10.10? According to Wikipedia, 10.8 ups the requirements for the system, but I don't think it is due to CPU features, so I don't buy into "CPU extensions (…) can be properly optimized".
As it may cut off some people, I'm fine with only doing it after the imminent release. I think @IntelOrca wanted this for next version as well.

**Comment by IntelOrca** [comment]
This will be merged after 0.0.7.
@LRFLEW @marijnvdwerf can you do what's necessary to make this build pass?

**Comment by janisozaur** [comment]
@marijnvdwerf @LRFLEW for what it's worth, openage have hit the same issue about `clock_gettime`: https://github.com/SFTtech/openage/issues/819

File: src/openrct2/platform/shared.c [hunk 2]
```diff
@@ -765,6 +775,15 @@ uint32 platform_get_ticks()
 {
 #ifdef _WIN32
     return GetTickCount();
+#elif defined(__APPLE__) && !defined(MACOS_SIERRA_ONWARDS)
+	mach_timebase_info_data_t mach_base_info;
+	kern_return_t ret = mach_timebase_info(&mach_base_info);
+	if (ret == 0){
```
Comment (janisozaur) [inline src/openrct2/platform/shared.c]: space missing, same at `else`

File: src/openrct2/platform/shared.c [hunk 3]
```diff
@@ -765,6 +775,15 @@ uint32 platform_get_ticks()
 {
 #ifdef _WIN32
     return GetTickCount();
+#elif defined(__APPLE__) && !defined(MACOS_SIERRA_ONWARDS)
+	mach_timebase_info_data_t mach_base_info;
+	kern_return_t ret = mach_timebase_info(&mach_base_info);
```
Comment (janisozaur) [inline src/openrct2/platform/shared.c]: This looks like something that should be cached

Comment (rwjuk) [inline src/openrct2/platform/shared.c]: True. Will do.

File: src/openrct2/platform/shared.c [hunk 4]
```diff
@@ -776,14 +781,15 @@ uint32 platform_get_ticks()
...
-	} else {
-		log_fatal("Unable to get mach_timebase_info.");
-		exit(-1);
+	if (!mach_info_initialised) {
+		kern_return_t ret = mach_timebase_info(&mach_base_info);
+		if (ret != 0) {
+			log_fatal("Unable to get mach_timebase_info.");
+			exit(-1);
+		}
+		mach_info_initialised = true;
```
Comment (janisozaur) [inline src/openrct2/platform/shared.c]: I'd rather see that in `platform_init` or `core_init`.

Compiler cannot optimise away access to globals, especially not ones that aren't marked `static` and this function can be quite hot.

Comment (rwjuk) [inline src/openrct2/platform/shared.c]: Makes sense - trouble is that if I put it in `platform_init`or `core_init` I end up with a division-by-zero error, as `platform_get_ticks()`is called during network initialisation, which (evidently!) happens before either of those functions are called.

Comment (IntelOrca) [inline src/openrct2/platform/shared.c]: It should go in `core_init`. It is one of the very first methods that is called - deliberately created for this kind of setup. I just tested it and it gets called before `platform_get_ticks`.

Comment (rwjuk) [inline src/openrct2/platform/shared.c]: Yeah, I was being an idiot before and assuming things. Moved to `core_init()``

File: src/openrct2/platform/shared.c [hunk 5]
```diff
@@ -822,4 +808,12 @@ uint8 platform_get_currency_value(const char *currCode) {
 void core_init()
 {
 	bitcount_init();
+	
+#if defined(__APPLE__) && !defined(MACOS_SIERRA_ONWARDS)
```
Comment (janisozaur) [inline src/openrct2/platform/shared.c]: you removed `MACOS_SIERRA_ONWARDS` in this commit

Comment (rwjuk) [inline src/openrct2/platform/shared.c]: Whoops! Clearly doing that stuff late at night was a bad move. Thanks for fixing!

**Comment by marijnvdwerf** [comment]
```
dyld: lazy symbol binding failed: Symbol not found: _clock_gettime
  Referenced from: /Users/Marijn/Library/Developer/Xcode/DerivedData/OpenRCT2-dprqycfasycyyffyzrmvaramdpki/Build/Products/Debug/OpenRCT2.app/Contents/MacOS/OpenRCT2
  Expected in: /usr/lib/libSystem.B.dylib

dyld: Symbol not found: _clock_gettime
  Referenced from: /Users/Marijn/Library/Developer/Xcode/DerivedData/OpenRCT2-dprqycfasycyyffyzrmvaramdpki/Build/Products/Debug/OpenRCT2.app/Contents/MacOS/OpenRCT2
  Expected in: /usr/lib/libSystem.B.dylib

/Users/Marijn/Library/Developer/Xcode/DerivedData/OpenRCT2-dprqycfasycyyffyzrmvaramdpki/Build/Intermediates/OpenRCT2.build/Debug/OpenRCT2.build/Script-D42C09D21C254F4E00309751.sh: line 2: 13590 Trace/BPT trap: 5       "${TARGET_BUILD_DIR}/${EXECUTABLE_PATH}" sprite build "${SRCROOT}/data/g2.dat" "${SRCROOT}/resources/g2/"
Command /bin/sh failed with exit code 133
```

**Comment by marijnvdwerf** [comment]
`core_init` gets called after `platform_get_ticks`

```
#0	0x00000001001323c8 in platform_get_ticks at /Users/Marijn/Projects/OpenRCT2/Main/src/openrct2/platform/shared.c:783
#1	0x0000000100134599 in NetworkConnection::ResetLastPacketTime() at /Users/Marijn/Projects/OpenRCT2/Main/src/openrct2/network/NetworkConnection.cpp:153
#2	0x00000001001344f9 in NetworkConnection::NetworkConnection() at /Users/Marijn/Projects/OpenRCT2/Main/src/openrct2/network/NetworkConnection.cpp:33
#3	0x0000000100134605 in NetworkConnection::NetworkConnection() at /Users/Marijn/Projects/OpenRCT2/Main/src/openrct2/network/NetworkConnection.cpp:32
#4	0x0000000100173f8f in Network::Network() at /Users/Marijn/Projects/OpenRCT2/Main/src/openrct2/network/network.cpp:89
#5	0x00000001001736b5 in Network::Network() at /Users/Marijn/Projects/OpenRCT2/Main/src/openrct2/network/network.cpp:90
#6	0x00000001001a53c4 in ::__cxx_global_var_init() at /Users/Marijn/Projects/OpenRCT2/Main/src/openrct2/network/network.cpp:76
#7	0x00000001001a53f9 in _GLOBAL__sub_I_network.cpp ()
```

**Comment by janisozaur** [comment]
@marijnvdwerf I assume that travis now fails due to division by zero in the example you provided?

**Comment by marijnvdwerf** [comment]
Jup.
Op vr 5 mei 2017 om 10:49 schreef Michał Janiszewski <
notifications@github.com>

**Comment by janisozaur** [comment]
Can we just expose a C API from `core/` to `std::chrono::steady_clock`? Windows' `GetTickCount` already uses 10~16ms period as "tick" and this would save us from the issues we face now. We already do similar thing with `Stopwatch`.

**Comment by IntelOrca** [comment]
`core_init` should be called first. If `platform_get_ticks` is being called from static initialisation, then it needs to be fixed so that it doesn't.

**Comment by janisozaur** [comment]
The easiest way, if this is the only occurence, would be to change `Network::server_connection` to a pointer and only initialise it in `Network::Init()`.

File: src/openrct2/core/FileStream.hpp [hunk 6]
```diff
@@ -123,25 +130,51 @@ class FileStream final : public IStream
 
     void Seek(sint64 offset, sint32 origin) override
     {
+#if defined(_MSC_VER)
         switch (origin) {
         case STREAM_SEEK_BEGIN:
-            SDL_RWseek(_file, offset, RW_SEEK_SET);
+            _fseeki64(_file, offset, SEEK_SET);
```
Comment (marijnvdwerf) [inline src/openrct2/core/FileStream.hpp]: What about conditionally defining `fseek64` to either `_fseeki64`, `fseeko` or`fseeko64`?

File: src/openrct2/network/network.cpp [hunk 7]
```diff
@@ -135,6 +133,7 @@ bool Network::Init()
 
 	status = NETWORK_STATUS_READY;
 
+	server_connection = new NetworkConnection();
```
Comment (marijnvdwerf) [inline src/openrct2/network/network.cpp]: Is this the proper casing for private class members?

File: src/openrct2/platform/shared.c [hunk 8]
```diff
@@ -761,7 +775,23 @@ void platform_set_cursor_position(sint32 x, sint32 y)
 
 uint32 platform_get_ticks()
 {
-	return SDL_GetTicks();
+#ifdef _WIN32
```
Comment (marijnvdwerf) [inline src/openrct2/platform/shared.c]: Shouldn't these be in the seperate platform files? Not really `shared` if there are so much platform checks

Comment (IntelOrca) [inline src/openrct2/platform/shared.c]: This will all be going anyway after my SDL-less refactor.

File: src/openrct2/network/network.cpp (hunk 7 above)
Comment (IntelOrca) [inline src/openrct2/network/network.cpp]: No, but it should be changed in a new PR.

**Comment by marijnvdwerf** [comment]
It runs for me now

**Comment by janisozaur** [comment]
My latest commit solves the issue I reported about chat messages being permanently visible.

**Pull request merged**